### ISSUE TITLE
[2.3.7] Script injection vulnerability in Javascript Router Generator

### ISSUE THREAD

**Issue by Lennholm** [playframework/playframework#4513]
Play version: 2.3.7 (Scala)

Thanks to a Whitehat security scan, I recently discovered a vulnerability in the Javascript Router Generator where it's possible for a client to inject code that is then executed when the webpage downloads the generated jsroutes file.

In the generated file we can find the following line of code:

``var _wA = function(r){return {ajax:function(c){c=c||{};c.url=r.url;c.type=r.method;return jQuery.ajax(c)}, method:r.method,type:r.method,url:r.url,absoluteURL: function(s){return _s('http',s)+'example.domain.com'+r.url},webSocketURL: function(s){return _s('ws',s)+'example.domain.com'+r.url}}}``

The above code creates the functions that generates the absolute URL and web socket URL of the route and the ``example.domain.com`` string is taken straight from the host header of the request.

By writing a different host header, the absolute URL can be made to point to a completely different URL which will cause files outside of your control to be loaded, or the file to not be loaded at all, depending on how the client handles cross-domain requests.
Alternatively, it can be made to point all requests to another resource on your system which an attacker can use to mess things up.

Although these absolute URL's aren't normally used, this isn't even the most dangerous aspect of this vulnerability.

It turns out the content of the host header isn't sanitized in any way which means it can be used to inject any code. For instance, writing the following to the host header:

``'}}};alert(1);a={a:{a:{a:'``

...will result in:

``var _wA = function(r){return {ajax:function(c){c=c||{};c.url=r.url;c.type=r.method;return jQuery.ajax(c)}, method:r.method,type:r.method,url:r.url,absoluteURL: function(s){return _s('http',s)+''}}};alert(1);a={a:{a:{a:''+r.url},webSocketURL: function(s){return _s('ws',s)+''}}};alert(1);a={a:{a:{a:''+r.url}}}``

where the ``alert(1)`` part is executed immediately as the client parses the file.

This can potentially be used by a compromised client or man-in-the-middle to compromise your system.

On the web site I'm working on, I worked around this issue by creating a new implicit request with the host header always set to ``'+window.location.host+'`` before calling the Javascript router generator.

**Comment by jroper** [comment]
When you say "compromised" client, what do you mean by compromised?  Ie, does the client already need to have had an XSS vulnerability exploited?  If so, doesn't this XSS vulnerability require having already exploited an XSS vulnerability, thus this would achieve nothing?  Similarly for a man in the middle, if the man in the middle can modify the host header, then the man in the middle can modify anything about the request, making exploiting an XSS vulnerability uninteresting.

Not saying this shouldn't be fixed, I just want to understand whether we have an actual security vulnerability here, and whether we need an advisory, backports etc.

**Comment by Lennholm** [comment]
It could be compromised in any way really, not necessarily XSS. But your points are all valid, if a client is affected by this they likely have much bigger problems. 
I'm no security expert by any means so it would probably be better if someone who is could comment on this aspect.
It's good to fix it though, putting something provided by the client into the response without any validation or escaping is never a good idea.


I should note that the proposed pull request doesn't fix the thing that led me to discover this, namely that the Whitehat security scan was able to insert its own domain as the host, for which it triggered a warning as a potential threat.
Of course, there's no reliable way for the application to know how the client connected to the server and the host header is used, by the client, to convey this. Perhaps this should be documented so that application developers can implement their own validation of the host header if they use Whitehat security (or similar tools) and run into this issue.

**Comment by gmethvin** [comment]
@Lennholm You are correct. The PR just escapes the host in the JavaScript which is a good idea but it doesn't prevent the potential issues caused by forging the host. If there is any caching going on in between in between the client and your Play server, your application might be vulnerable to cache poisoning attacks (see http://www.skeletonscribe.net/2013/05/practical-http-host-header-attacks.html and https://www.owasp.org/index.php/Cache_Poisoning).

The preferred solution seems to be to validate the host header against a list of allowed hosts. [This is what Django does](https://docs.djangoproject.com/en/1.8/topics/security/#host-header-validation), for example. So we could possibly add an allowed hosts configuration key that Play checks when getting the `host` property of the request.

**Comment by richdougherty** [comment]
Would it work to get that information from the DOM instead of from the request headers, e.g. using [`window.location.hostname`](https://developer.mozilla.org/en-US/docs/Web/API/Location) or something like that?

**Comment by jroper** [comment]
Getting it from the DOM won't work because the host won't necessarily be the same one that the browser is currently at - if I've got an API that uses CORS, I might expose a JavaScript router for invoking that API, in which case, I would need to ensure that the absolute URL was for my host, and not the one that the browser is currently at.

**Comment by jroper** [comment]
Solving this in Play I think would best be done by adding a filter that validates the host header against a list of trusted hosts:

```scala
class SecureHostFilter(allowedHosts: Set[String], allowNone: Boolean, errorHandler: HttpErrorHandler) extends EssentialFilter {
  def apply(next: EssentialAction) = EssentialAction { req =>
    req.host match {
      case allowed if allowedHosts(allowed) => next(req)
      case "" if allowNone => next(req)
      case blocked => Iteratee.ignore.mapM(_ => errorHandler.onClientError(400, s"Disallowed/forged host header: $blocked"))
    }
  }
}
```

**Comment by gmethvin** [comment]
A filter seems like a good approach, similar to how we do it with CSRFFilter. I do think it'd be fairly easy to do this on your own, but it might be nice to ship the filter with Play and provide a consistent configuration format. I'd ideally like to see something where you could specify `*.example.com.`  and it'd match `foo.example.com`, `bar.example.com`, `foo.example.com.` (with trailing dot), etc.

**Comment by Lennholm** [comment]
@gmethvin I experimented a bit with a whitelist solution before I settled on always overriding the header with a safe value, which works well for my project. It turned ugly when I had to create different whitelists depending on which environment the application was running in (dev, stage or prod) but putting this as a configuration could alleviate that.

@richdougherty That's essentially how I ended up working around this issue, by overriding the host header with ``'+window.location.host+'`` the jsroutes file will contain the code ``_s('http',s)+''+window.location.host+''+r.url``.
I'm not sure this is a catch-all solution though, what if the client isn't meant to be a browser but a proprietary front-end application where ``window`` isn't a thing? Then again, the js router appears to (by default) assume that jQuery is running in the client, so does Play only support a front-end that runs in browsers?
